DEVELOPING_TUTORIALS.md

```diff
@@ -603,6 +614,133 @@ The properties exported from this file depend on the lesson type, as follows:
   - `question`: String - mandatory - _the question to be answered by the user_ ([detailed docs](#create-multiple-choice-quizzes-in-your-javascript-file-skip-for-coding-exercises-and-text-only-lessons))
   - `choices`: Array - mandatory - _the potential answers between which the user must choose_ ([detailed docs](#create-multiple-choice-quizzes-in-your-javascript-file-skip-for-coding-exercises-and-text-only-lessons))
 
+### `utils` module
+
+This module is a set of utils designed to be re-used accross the validation code of different tutorials.
+
+#### `utils.format`
+
+- `ipfsObject(object)`: formats an IPFS object to be ready to print in the output UI, for example, formatting a `cid` object into a string.
+
+Example:
+
+```js
+return {
+  success: 'Success! You did it!',
+  logDesc: "Here is the result.",
+  log: utils.format.ipfsObject(result)
+}
+```
+
+Original `result` object would show like this:
+
+```
+{
+  "path": "QmQNcG7yJPWGuAoM2jmF78t6GTdNSmcuXr2c18SfJ4tVJQ",
+  "cid": {
+    "codec": "dag-pb",
+    "version": 0,
+    "hash": {
+      "type": "Buffer",
+      "data": [
+        18,
+        32,
+        30,
+        55,
+        196,
+        121,
+        86,
+        111,
+        123,
+        58,
+        178,
+        22,
+        159,
+        46,
+        21,
+        223,
+        12,
+        1,
+        37
+      ]
+    }
+  },
+  "size": 40391,
+  "mode": 420
+}
+```
+
+While the formatted object would show like this:
+
+```json
+{
+  "path": "QmRz7dg5FcbESqbbzxMVksyHfqJwqibcKZULUoichwwmsu",
+  "cid": "CID('QmRz7dg5FcbESqbbzxMVksyHfqJwqibcKZULUoichwwmsu')",
+  "size": 778308,
+  "mode": 420
+}
+```
+
+#### `utils.ipfs`
+
+- `errorCodes`: set of error codes from `js-ipfs`
+  - `ERR_MORE_THAN_ONE_ROOT`
+
+```js
+if (result.code === utils.ipfs.errorCodes.ERR_MORE_THAN_ONE_ROOT) {
+  return {
+    fail: 'Some error message specific to this edge case.'
+  }
+}
+```
+
+
+
+#### `utils.validationMessages`
+
+These validation messages can be used to avoid writing the same message for the users across tutorials and lessons.
+
+- `SUCCESS`: when the user has successfully completed the code challenge
+  - Example: `Success! You did it!`
+- `NO_RESULT`: to show when the user forgets to return a result
+  - Example: `Oops! You forgot to return a result :(`
+- `VALUE_IS_ASYNC_ITERABLE_ALL`: when the user returned an AsyncIterable and forgot to use the `all` function on the result
+  - Example: `The returned value is an Async Iterable. Did you forget to put all the results together using either for await...of or with all?`
+- `VALUE_IS_ASYNC_ITERABLE_CONCAT`: when the user returned an AsyncIterable and forgot to use the `concat` function on the result
+  - Example: `The returned value is an Async Iterable. Did you forget to concatenate all the data using concat?`
+
+
```

Comment:
While these codes provide an easy shorthand, you'll still need to provide the logic that determines when a particular validation message should be displayed for each lesson. For example:

Reply:
```suggestion
While these codes provide an easy shorthand, you'll still need to provide the logic that determines when a particular validation message should be displayed for each lesson. For example:
```